README.md
# Styleguide Gent Base

This style guide contains the SASS library and needed assets to be able to style
a web application in the corporate identity for the city of Ghent..

[![NPM version](https://img.shields.io/npm/v/gent_styleguide.svg)](https://www.npmjs.com/package/gent_styleguide)
[![Downloads](https://img.shields.io/npm/dt/gent_styleguide.svg)](https://www.npmjs.com/package/gent_styleguide)
[![License](https://img.shields.io/github/license/StadGent/fractal_styleguide_gent-base.svg)](https://www.npmjs.com/package/gent_styleguide)

[![Build Status](https://travis-ci.org/StadGent/fractal_styleguide_gent-base.svg?branch=develop)](https://travis-ci.org/StadGent/fractal_styleguide_gent-base)
[![Maintainability](https://api.codeclimate.com/v1/badges/1aeba4d73f307d4583b9/maintainability)](https://codeclimate.com/github/StadGent/fractal_styleguide_gent-base/maintainability)

## Roadmap

### Style guide version 3.x

* A major overhaul with an entirely new style for all components.
* Probably a full CSS grid implementation throughout all the style guide
components.
* We will probably drop jQuery in favor for native Javascript or another.

### Upcoming: 2.7.x

* General changes and fixes to the style guide.

## Licenses

If you are building something for the City of Ghent (Stad Gent) or Digipolis,
you can contact the web team at Digipolis or open an issue here to get the
necessary licenses thar are required for your project.
For more info on what licenses you need see below.

### License: Expressway font

The Expressway font is a paying font used by the style guide of the City of
Ghent. It needs to be included through Typekit.

### License: Light Gallery plugin for jQuery

It is important to note that anyone that wants to use this style guide or fork
this repository for commercial purposes, should [pay for a license](http://sachinchoolur.github.io/lightGallery/docs/license.html)!

## Folder structure

When installing this package, you will get a directory structure like this.

```
.
├── CHANGELOG.md
├── LICENSE
├── README.md
├── build
│   └── styleguide
│       └── sass
└── package.json
```

The build directory contains all SASS source files and all asset files needed to
be able to style a web application in the corporate identity for the city of
Ghent..

## Implementing this style guide in your own project

This style guide aims to be technology independent. This means you should be
able to use it inside your project with any technology you want.
It utilizes SASS to style its components, so you will have to use that in your
project.

### Setup

After installing the style guide through `npm install gent_styleguide` you''ll
get a folder structure like this:

```
.
├── node_modules
│   └── gent_styleguide
│   └── breakpoint-sass
├── package-lock.json
└── package.json
```

Now, to use the style guide SASS partials you just need to import them into your
main SASS file at the top so they get loaded before your own code.

```
@import 'node_modules/gent_styleguide/build/sass/00-settings/reset';
@import 'node_modules/gent_styleguide/build/sass/00-settings/vars';
@import 'node_modules/gent_styleguide/build/sass/01-mixins/**/*';
@import 'node_modules/gent_styleguide/build/sass/11-base/**/*';
@import 'node_modules/gent_styleguide/build/sass/21-atoms/**/*';
@import 'node_modules/gent_styleguide/build/sass/31-molecules/**/*';
@import 'node_modules/gent_styleguide/build/sass/41-organisms/**/*';
```

Note: we use this structure to allow you to override anything you wish in your
own custom project.
The way to do this, is to add the imports above to your main SASS file and add
imports of your own overriding files in between.

If you don't need to be able to change the molecules for your project and
instead just want to import the style guide
and use it as it is, you can use the `main.scss` file instead of using the
following approach.

Example:

```
@import 'node_modules/gent_styleguide/build/sass/00-settings/reset';
@import 'node_modules/gent_styleguide/build/sass/00-settings/vars';
@import '/sass/00-vars-overrides/**/*';
@import 'node_modules/gent_styleguide/build/sass/01-mixins/**/*';
@import '/sass/01-mixins-overrides/**/*';
@import 'node_modules/gent_styleguide/build/sass/11-base/**/*';
@import '/sass/11-base-overrides/**/*';
@import 'node_modules/gent_styleguide/build/sass/21-atoms/**/*';
@import '/sass/21-atom-overrides/**/*';
@import 'node_modules/gent_styleguide/build/sass/31-molecules/**/*';
@import '/sass/31-molecule-overrides/**/*';
@import 'node_modules/gent_styleguide/build/sass/41-organisms/**/*';
@import '/sass/41-organism-overrides/**/*';
```

Note: When using the style guide you will need to add breakpoint-sass as an
includePath inside your gulpfile.js

Example:

```
    .pipe(sass({
      outputStyle: 'compressed',
      includePaths: ['node_modules/breakpoint-sass/stylesheets']
    })).on('error', sass.logError)
```

This includePaths path may vary depending on your setup!

### Fonts

You need to make sure the Expressway font is included in your project somehow.
This would be done by including the Typescript script in your project and
requesting a license / Typekit kit from the web team at Digipolis.

Example:

```
<script>
  (function(d) {
    var config = {
          kitId: 'kgt4wbp',
          scriptTimeout: 3000,
          async: true
        },
        h=d.documentElement,t=setTimeout(function(){h.className=h.className.replace(/\bwf-loading\b/g,"")+" wf-inactive";},config.scriptTimeout),tk=d.createElement("script"),f=false,s=d.getElementsByTagName("script")[0],a;h.className+=" wf-loading";tk.src='https://use.typekit.net/'+config.kitId+'.js';tk.async=true;tk.onload=tk.onreadystatechange=function(){a=this.readyState;if(f||a&&a!="complete"&&a!="loaded")return;f=true;clearTimeout(t);try{Typekit.load(config)}catch(e){}};s.parentNode.insertBefore(tk,s)
  })(document);
</script>
```

So in this example you would need to change the `kitId` to the ID you got from
the web team of Digipolis.

## Third party support

### AngularJS CLI support (no SASS globbing)

We provide a `main_cli.scss` file that is essentially the same as `main.scss`,
but includes all SASS partials the style guide needs without SASS globbing.

#### IMPORTANT

We recommend using the `main_cli.scss` file only when you need to support
Angular CLI.


CHANGELOG.md
# Changelog

## Roadmap

Refer to upcoming changes in our README.md under "Roadmap"

## Changes

All notable changes to `digipolisgent/fractal_styleguide_gent-base`.

### 2.9.0

* STIJ-246: Added a mixin  and class `no-link-style` to make it possible
  to override the link styling of certain blocks with `text-decoration: none;`.
* STIJ-246: Give links that are styled like a button a `text-decoration: none;`.
* STIJ-86: Added documentation for the description atom.
* STIJ-93: Added external link variant. This defines a class `.external-link`
  that provides a `::before` on the link and adds a new-tab icon.
* STIJ-92: Updated the label atom documentation.
* STIJ-88: Updated the heading atom documentation.
* STIJ-96: Updated the paragraph atom documentation.
* STIJ-94: Updated the list atom documentation.
* STIJ-250: Added small fix for empty teaser collection preview templates.

### 2.8.1

* STIJ-244: Fixed critical gulp build bug. **IMPORTANT** Necessary fix for 2.8.0.

### 2.8.0

* STIJ-198: Changed color-element-states() mixin to element-states() mixin.
* STIJ-236: Fixed markup of breadcrumb + fixed an issue where the
  breadcrumb didn't wrap correctly.
* STIJ-237: **IMPORTANT** Removed bootstrap grid styling for forms.

### 2.7.8

* STIJ-200: Refactor status messages.
* STIJ-229 Refactored Gulp-file to ECMA-6
* STIJ-229 Fixed gulp npm-install. Added --help support and argument checking
  through the yargs module.
* STIJ-229 Fixed gulp build to support -h --help, option validation through Yargs.
* STIJ-229 Refactored ES-lint to accomodate node.js, ES-6, enforced
  'use strict'; on a function level instead of global.

### 2.7.7

* STIJ-235: Refactor form actions.

### 2.7.6

* STIJ-221: Removed header margin top again. We need to think of a better way to
  implement this.
* STIJ-219: Added travis integration.
* STIJ-223: Added codeclimate integration.

### 2.7.5

* STIJ-216:
  * Added masonry-layout library.
  * Added organism for masonry layout.
* STIJ-217: Added validation check to version_tag.sh script.

### 2.7.4

* STIJ-197: Added margin top to headings and fixed the color of links inside
  headings.
* STIJ-207: Fixed small issue for preview templates.
* STIJ-208: Fixed bug in `postinstall.sh` script.
* STIJ-210: Move some dependencies to devDependencies.
* STIJ-211: Fixed newsletter block styling.
* STIJ-214: Added preview template fixes after refactoring the heading atoms.

### 2.7.3

Due to some issues with publishing to NPM we had to update to a new patch
version without introducing new changes to the style guide.

### 2.7.2

* STIJ-208: Removed console log and updated postinstall.sh script to generate
  main_cli.scss content.

### 2.7.1

* STIJ-170: Fixed a bug where the search icon in the header did not work for FF
  and IE.
* STIJ-175: Added teaser and topic teaser molecules.
* STIJ-193: Added figcaption atom and figure molecule.
* STIJ-194: Updated preview layouts and added some building blocks.
* STIJ-199: Removed some style guide specific preview templates that where not
  necessary anymore.
* STIJ-205: Added deploy script for admins.
* STIJ-206: Added main_cli.scss file that creates the main sass file **without
  SASS globbing** to support Angular CLI.
* STIJ-207: Added small fix for preview layouts.

### 2.7.0

IMPORTANT: Class .text-center has been removed

* STIJ-1: Fixed whitespace issue for the gallery component.
* STIJ-30: Fixed styling for select and multi select fields on iPad.
* STIJ-39: Fixed component colors inside of sections, based on guidelines fron
  the City of Ghent.
* STIJ-45: Changed jQuery UI datepicker format to a Dutch format.
* STIJ-87: Added description field documentation.
* STIJ-87: Refactored description field atom to use Themify mixin.
* STIJ-89: Changed the documentation of an icon to English and update it a bit.
* STIJ-96: Added paragraph atom documentation.
* STIJ-96: Removed the paragraph variant that contains centered text, **it is
  not allowed by the City of Ghent**!
* STIJ-100: Added variant for CTA links.
* STIJ-100: Added CTA link documentation.
* STIJ-106: Added search molecule documentation.
* STIJ-170: Fixed search molecule bug where it was using the old button submit
  atom instead of the input type submit atom.
* STIJ-170: Added a label to the search molecule.
* STIJ-170: Hide the label of the search molecule in the header organism.
* STIJ-170: Add placeholder to search molecule inside the header organism.
* STIJ-175: Added page wide image component.
* STIJ-177: Added newsletter block component.
* STIJ-178: Added styling for inline form-items.
* STIJ-178: Added address organism.
* STIJ-182: Fixed whitespace when previewing footer component in the style guide.
* STIJ-186: Added header variant with breadcrumb and with a banner image.
* STIJ-186: Added small fix to the buttons in the header.
* STIJ-188: Fixed issue in gulp command js:dist.
* STIJ-189: Fixed some style guide specific issues.

### 2.6.19

* STIJ-158: Changed datepicker date format to an English format.
* STIJ-159: Removed button submit atom.
* STIJ-159: Moved the submit button documentation to the input submit atom.
* STIJ-160: Changed the form actions - editor molecule variant to include a
  normal button instead of a dropbutton.
* STIJ-161: Fixed broken links in documentation files.
* STIJ-162: Changed all Dutch text to English.
* STIJ-168: Changed style guide title to a versioned version.
* STIJ-169: Integrate link in style guide to the GitHub CHANGELOG.

### 2.6.18

* STIJ-50: Fixed label for attributes and changed a lot of the form examples in
  the style guide.
* STIJ-86: Added documentation to some components.
* STIJ-156: Updated the CONTRIBUTING and README files.
* STIJ-156: Updated the README file of the gallery component with important
  licensing information.
* STIJ-157: Changed the loaded JS files in the _preview.twig file to the
  minified versions. This fixes issues when automatically deploying.

gulpfile.js
'use strict';

const packageInfo = require('./package.json');
const packageVersion = packageInfo.version;

/*
* Node core modules.
*/
const fs = require('fs');
const path = require('path');

/*
* NPM based modules
*/
const gulp = require('gulp');
const sass = require('gulp-sass');
const sassGlob = require('gulp-sass-glob');
const sourcemaps = require('gulp-sourcemaps');
const sassLint = require('gulp-sass-lint');
const autoprefixer = require('gulp-autoprefixer');
const cssnano = require('gulp-cssnano');
const rename = require('gulp-rename');
const eslint = require('gulp-eslint');
const imagemin = require('gulp-imagemin');
const pngquant = require('imagemin-pngquant');
const minify = require('gulp-minify');
const npm = require('npm');
const bump = require('gulp-bump');
const inject = require('gulp-inject');
const yargs = require('yargs');

var _sassLint = (failOnError) => {
  var cmd = gulp.src('components/**/*.s+(a|c)ss')
    .pipe(sassGlob())
    .pipe(sassLint({
      configFile: './.sass-lint.yml'
    }))
    .pipe(sassLint.format());

  if (failOnError) {
    cmd.pipe(sassLint.failOnError());
  }

  return cmd;
};

/*
* Require the Fractal module
*/
const fractal = require('@frctl/fractal').create();
const logger = fractal.cli.console; // keep a reference to the fractal CLI console utility

/**
 * Require additional fractal modules
 */
const twigAdapter = require('@frctl/twig');

/*
* Give your project a title.
*/
fractal.set('project.title', 'City of Ghent Style Guide - Version ' + packageVersion);

/*
* Tell Fractal where to look for components.
*/
fractal.components.set('path', path.join(__dirname, 'components'));
fractal.components.set('default.preview', '@preview');
fractal.components.engine(twigAdapter);
fractal.components.set('ext', '.twig');

/*
* Set custom statuses.
*/
fractal.components.set('statuses', {
  deprecated: {
    label: 'deprecated',
    description: 'Deprecated.',
    color: '#dd5e01'
  },
  beta: {
    label: 'beta',
    description: 'Work in progress. Implement with caution.',
    color: '#ff9233'
  },
  ready: {
    label: 'Ready',
    description: 'Ready to implement.',
    color: '#29cc29'
  }
});

/*
* Tell Fractal where to look for documentation pages.
*/
fractal.docs.set('path', path.join(__dirname, 'docs'));

/*
* Tell the Fractal web preview plugin where to look for static assets.
*/
fractal.web.set('static.path', path.join(__dirname, 'public'));
fractal.web.set('static.mount', '');
fractal.web.set('builder.dest', __dirname + '/build');

/*
 *
 * Inject SASS partial paths as imports in main_cli.scss.
 *
 */
gulp.task('styles:inject', () => {
  const injectSettingsFiles = gulp.src('components/00-settings/**/*.s+(a|c)ss', {read: false});
  const injectMixinsFiles = gulp.src('components/01-mixins/**/*.s+(a|c)ss', {read: false});
  const injectBaseFiles = gulp.src('components/11-base/**/*.s+(a|c)ss', {read: false});
  const injectAtomsFiles = gulp.src('components/21-atoms/**/*.s+(a|c)ss', {read: false});
  const injectMoleculesFiles = gulp.src('components/31-molecules/**/*.s+(a|c)ss', {read: false});
  const injectOrganismsFiles = gulp.src('components/41-organisms/**/*.s+(a|c)ss', {read: false});

  var transformFilepath = (filepath) => `@import "${filepath}";`;

  const injectSettingsOptions = {
    transform: transformFilepath,
    starttag: '// inject:settings',
    endtag: '// endinject',
    addRootSlash: false,
    relative: true
  };

  const injectMixinsOptions = {
    transform: transformFilepath,
    starttag: '// inject:mixins',
    endtag: '// endinject',
    addRootSlash: false,
    relative: true
  };

  const injectBaseOptions = {
    transform: transformFilepath,
    starttag: '// inject:base',
    endtag: '// endinject',
    addRootSlash: false,
    relative: true
  };

  const injectAtomsOptions = {
    transform: transformFilepath,
    starttag: '// inject:atoms',
    endtag: '// endinject',
    addRootSlash: false,
    relative: true
  };

  const injectMoleculesOptions = {
    transform: transformFilepath,
    starttag: '// inject:molecules',
    endtag: '// endinject',
    addRootSlash: false,
    relative: true
  };

  const injectOrganismsOptions = {
    transform: transformFilepath,
    starttag: '// inject:organisms',
    endtag: '// endinject',
    addRootSlash: false,
    relative: true
  };

  return gulp.src('components/main_cli.scss')
    .pipe(inject(injectSettingsFiles, injectSettingsOptions))
    .pipe(inject(injectMixinsFiles, injectMixinsOptions))
    .pipe(inject(injectBaseFiles, injectBaseOptions))
    .pipe(inject(injectAtomsFiles, injectAtomsOptions))
    .pipe(inject(injectMoleculesFiles, injectMoleculesOptions))
    .pipe(inject(injectOrganismsFiles, injectOrganismsOptions))
    .pipe(gulp.dest('components/'));
});

/*
 *
 * Development settings of your styles.
 * Includes:
 *  Sass globbing
 *  SCSS linting
 *  Nested output style
 *  Sourcemaps (dev only!)
 *  Autoprefixer
 */
gulp.task('styles:dist', (callback) => {
  _sassLint(false)
    .pipe(sourcemaps.init())
    .pipe(sass({
      outputStyle: 'nested',
      includePaths: [
        'node_modules/breakpoint-sass/stylesheets',
        'node_modules/susy/sass'
      ]
    })).on('error', sass.logError)
    .pipe(autoprefixer({
      browsers: ['last 5 versions']
    }))
    .pipe(sourcemaps.write())
    .pipe(gulp.dest('./public/css/'));
  callback();
});

/*
 *
 * Build settings for your styles.
 * Includes:
 *  Sass globbing
 *  SCSS linting
 *  Compresssed output style
 *  Autoprefixer
 *
 */
gulp.task('styles:build', ['styles:inject', 'fractal:build'], (callback) => {
  _sassLint(true)
    .pipe(sass({
      outputStyle: 'compressed',
      includePaths: [
        'node_modules/breakpoint-sass/stylesheets',
        'node_modules/susy/sass'
      ]
    })).on('error', sass.logError)
    .pipe(autoprefixer({
      browsers: ['last 5 versions']
    }))
    .pipe(gulp.dest('./build/css/'))
    .pipe(cssnano())
    .pipe(gulp.dest('./build/css/'));
  callback();
});

/*
 *
 * Validate SCSS files.
 *
 */
gulp.task('styles:validate', (callback) => {
  _sassLint(true);
  callback();
});

/*
 *
 * Watch SCSS files For Changes.
 *
 */
gulp.task('styles:watch', (callback) => {
  gulp.watch('./components/**/*.scss', ['styles:dist']);
  callback();
});

/*
 *
 * Extract SCSS and their assets (like fonts) from the components folder.
 *
 */
gulp.task('styles:extract', [
  'fractal:build',
  'styles:build',
  'styles:dist'
], (callback) => {
  gulp.src('components/**/*.s+(a|c)ss')
    .pipe(gulp.dest('./build/styleguide/sass/'));
  callback();
});

/*
 *
 * Copy JS files during development.
 *
 */
gulp.task('js:dist', ['styles:dist'], (callback) => {
  gulp.src('components/**/*.js')
    .pipe(rename({
      dirname: '',
      suffix: '-min'
    }))
    .pipe(gulp.dest('./public/styleguide/js/'));
  callback();
});

/*
 *
 * Copy JS files during Fractal build.
 *
 */
gulp.task('js:build', ['fractal:build'], (callback) => {
  gulp.src('components/**/*.js')
    .pipe(rename({dirname: ''}))
    .pipe(minify({
      noSource: true
    }))
    .pipe(gulp.dest('./build/styleguide/js/'));
  callback();
});

/*
 *
 * Validate JS files.
 *
 */
gulp.task('js:validate', (callback) => {
  gulp.src('components/**/*.js')
    .pipe(eslint({
      configFile: './.eslintrc'
    }))
    .pipe(eslint.format())
    .pipe(eslint.failAfterError());

  callback();
});

/*
 *
 * Watch JS files For Changes.
 *
 */
gulp.task('js:watch', (callback) => {
  gulp.watch('./components/**/*.js', ['js:validate', 'js:dist']);
  callback();
});

/*
 *
 * Minify images.
 *
 */
gulp.task('images:minify', [
  'fractal:build',
  'styles:build',
  'styles:dist'
], (cb) =>
  gulp.src([
    'components/**/*.png',
    'components/**/*.jpg',
    'components/**/*.gif',
    'components/**/*.jpeg',
    'components/**/*.svg'
  ])
    .pipe(imagemin({
      progressive: true,
      use: [pngquant()]
    })).pipe(gulp.dest('build/styleguide/sass'))
);

/*
 * Start the Fractal server
 *
 * In this example we are passing the option 'sync: true' which means that it will
 * use BrowserSync to watch for changes to the filesystem and refresh the browser automatically.
 * Obviously this is completely optional!
 *
 * This task will also log any errors to the console.
 */
gulp.task('fractal:start', () => {
  const server = fractal.web.server({
    sync: true
  });
  server.on('error', err => logger.error(err.message));
  return server.start().then(() => {
    logger.success(`Fractal server is now running at ${server.url}`);
  }).catch(() => logger.error('Fractal server failed to start'));
});

/*
 * Run a static export of the project web UI.
 *
 * This task will report on progress using the 'progress' event emitted by the
 * builder instance, and log any errors to the terminal.
 *
 * The build destination will be the directory specified in the 'builder.dest'
 * configuration option set above.
 */
gulp.task('fractal:build', () => {
  const builder = fractal.web.builder();
  // builder.on('progress', (completed, total) => logger.update(`Exported
  // ${completed} of ${total} items`, 'info'));
  builder.on('error', err => logger.error(err.message));
  return builder.build().then(() => {
    logger.success('Fractal build completed!');
  }).catch(() => logger.error('Fractal server failed to start'));
});

/*
 * Publish to the NPM public registry.
 */
gulp.task('publish:npm', (callback) => {

  const argv = yargs
    .options({
      username: {
        demand: true,
        alias: 'u',
        describe: 'NPM user name',
        string: true
      }
    })
    .options({
      password: {
        demand: true,
        alias: 'p',
        describe: 'NPM password',
        string: true
      }
    })
    .options({
      email: {
        demand: true,
        alias: 'e',
        describe: 'E-mail',
        string: true
      }
    })
    .help()
    .alias('help', 'h')
    .argv;


  const username = argv.username;
  const password = argv.password;
  const email = argv.email;

  const uri = 'http://registry.npmjs.org/';

  npm.load(null, (loadError) => {
    if (loadError) {
      return callback(loadError);
    }
    const auth = {
      username: username,
      password: password,
      email: email,
      alwaysAuth: true
    };
    const addUserParams = {
      auth: auth
    };

    npm.registry.adduser(uri, addUserParams, (addUserError, data, raw, res) => {
      if (addUserError) {
        return callback(addUserError);
      }
      let metadata = require('./package.json');
      metadata = JSON.parse(JSON.stringify(metadata));
      npm.commands.pack([], (packError) => { // eslint-disable-line max-nested-callbacks
        if (packError) {
          return callback(packError);
        }
        const fileName = metadata.name + '-' + metadata.version + '.tgz';
        const bodyPath = require.resolve('./' + fileName);
        const body = fs.createReadStream(bodyPath);
        const publishParams = {
          metadata: metadata,
          access: 'public',
          body: body,
          auth: auth
        };
        npm.registry.publish(uri, publishParams, (publishError, resp) => { // eslint-disable-line max-nested-callbacks
          if (publishError) {
            return callback(publishError);
          }
          callback(`Publish succesfull: ${JSON.stringify(resp, null, 2)}`);
          return callback();
        });
      });
    });
  });
});

/*
 * Bump the version number of the package.
 */
gulp.task('bump', () => {
  const argv = yargs
    .options({
      type: {
        demand: true,
        alias: 't',
        describe: 'NPM user name',
        string: true,
        choices: ['prerelease', 'patch', 'minor', 'major']
      }
    })
    .help()
    .alias('help', 'h')
    .argv;

  // Change version number of package.json file.
  gulp.src('./package.json')
    .pipe(bump({
      type: argv.type
    }))
    .pipe(gulp.dest('./'));
});

/*
 *
 * Default tasks:
 * Usage:
 *  gulp
 *  gulp watch
 *
 * Used for local development to compile and validate after every change.
 *
 */
gulp.task('default', ['fractal:start', 'styles:watch', 'js:watch']);
gulp.task('watch', ['default']);

/*
 *
 * Validate task:
 * Usage:
 *  gulp validate
 *
 *  Used to only validate the SCSS and JS code.
 *
 */
gulp.task('validate', [
  'styles:validate',
  'js:validate'
], callback => callback());

/*
 *
 * Compile task:
 * Usage:
 *  gulp compile
 *  gulp compile:dev
 *  Add sourcemaps to the CSS files.
 *
 *  Used to compile production ready SCSS and JS code.
 *
 */
gulp.task('compile', [
  'fractal:build',
  'styles:build',
  'styles:dist',
  'styles:extract',
  'js:build',
  'js:dist',
  'images:minify'
], callback => callback());
gulp.task('compile:dev', [
  'fractal:build',
  'styles:dist',
  'js:dist',
  'images:minify'
]);

/*
 *
 * Build task:
 * Usage:
 *  gulp build
 *
 *  Used to validate and build production ready code.
 *
 */
gulp.task('build', ['validate', 'compile'], () => {
  return gulp.src('components/**/*.s+(a|c)ss')
    .pipe(gulp.dest('./build/styleguide/sass/'))
});

/*
 *
 * Publish task:
 * Usage:
 *  gulp publish
 *
 *  Used to publish to the public NPM registry.
 *
 */
gulp.task('publish', ['publish:npm']);
